api/models/ml_model.py

```diff
@@ -0,0 +1,133 @@
+from pydantic import Field, validator, root_validator
+from typing import List, Optional
+
+from .base import SnakeModel
+
+
+class MLModelDTO(SnakeModel):
+    device: str = Field(example="Jetson")
+    name: str = Field(example="openvino")
+    image_size: str = Field(example="300,300,3")
+    model_path: str = Field("")
+    class_id: int = Field(1)
+    min_score: float = Field(0.25, example=0.30)
+    tensorrt_precision: Optional[int] = Field(None, example=32)
+
+    @validator("device")
+    def validate_device(cls, device):
+        if device not in ["x86", "x86-gpu", "Jetson", "EdgeTPU", "Dummy"]:
+            raise ValueError('Not valid Device. Try one of the following: "x86", "x86-gpu", "Jetson", "EdgeTPU",'
+                             '"Dummy".')
+        return device
+
+    @validator("tensorrt_precision")
+    def validate_tensorrt_precision(cls, tensorrt_precision):
+        if tensorrt_precision is not None:
+            if tensorrt_precision not in [16, 32]:
+                raise ValueError('Not valid tensorrt_precision. Accepted values: 16, 32')
+
+        return tensorrt_precision
+
+    @validator("name")
+    def validate_name(cls, name):
+        if name not in ["openvino", "openpifpaf_tensorrt", "mobilenet_ssd_v2", "openpifpaf", "yolov3",
+                              "ssd_mobilenet_v2_coco", "ssd_mobilenet_v2_pedestrian_softbio", "posenet",
+                              "pedestrian_ssd_mobilenet_v2", "pedestrian_ssdlite_mobilenet_v2"]:
+            raise ValueError('Not valid ML model. Try one of the following: "openvino", "openpifpaf_tensorrt",'
+                             '"mobilenet_ssd_v2", "openpifpaf", "yolov3",'
+                             '"ssd_mobilenet_v2_coco", "ssd_mobilenet_v2_pedestrian_softbio", "posenet",'
+                             '"pedestrian_ssd_mobilenet_v2", "pedestrian_ssdlite_mobilenet_v2".')
+        return name
+
+    @validator("image_size")
+    def validate_image_size(cls, image_size):
+        integers = image_size.split(",")
+
+        if len(integers) != 3:
+            raise ValueError('ImageSize must be a string with 3 numbers separated with commas. Ex: "30,30,40".')
+
+        for x in integers:
+            try:
+                int(x)
+            except ValueError:
+                raise ValueError('ImageSize must be a string with 3 numbers separated with commas. Ex: "30,30,40".')
+
+        return image_size
+
+    # Root validators are called after each field validators success.
+
+    @root_validator(skip_on_failure=True)
+    def check_models_and_device(cls, values):
```

Comment:
Please add all these strings in a dictionary as a constant

MODELS_DEVICES = { device: [models], device: [models]}

And refactor the validations to reuse information from there (so it is centralized). For example: in order to get all devices MODELS_DEVICES.keys()

We could do something similar with the variables in the validator below (by instead of storing an array of models we store a dictionary of variables, models being one of them), although maybe that's too complicated for images sizes that  are a multiple of something.

Reply:
Changed!